control-plane/agents/src/bin/core/controller/scheduling/volume_policy/node.rs

```diff
@@ -0,0 +1,126 @@
+        !registry
+            .specs()
+            .cordoned_nodes()
+            .into_iter()
+            .any(|node_spec| node_spec.id() == &item.pool.node)
+    }
+
+    /// Should only attempt to use online nodes.
+    pub(crate) fn online(_request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        item.node_wrapper().is_online()
+    }
+
+    /// Should only attempt to use nodes which are not cordoned.
+    pub(crate) fn cordoned(request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        let registry = request.registry();
+        !registry
+            .specs()
+            .cordoned_nodes()
+            .into_iter()
+            .any(|node_spec| node_spec.id() == item.node_wrapper().id())
+    }
+
+    /// Should only attempt to use node where current target is not present.
+    pub(crate) fn current_target(request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        if let Some(target) = request.target() {
+            target.node() != item.node_wrapper().id()
+        } else {
+            true
+        }
+    }
+    /// Should only attempt to use node where there are no targets for the current volume.
+    pub(crate) fn no_targets(request: &GetSuitableNodesContext, item: &NodeItem) -> bool {
+        let volume_targets = request.registry().specs().volume_nexuses(&request.uuid);
+        !volume_targets
+            .into_iter()
+            .any(|n| &n.lock().node == item.node_wrapper().id())
+    }
+    /// Should only attempt to use nodes having specific creation label if topology has it.
+    pub(crate) fn topology(request: &GetSuitablePoolsContext, item: &PoolItem) -> bool {
+        let volume_node_topology_inclusion_labels: HashMap<String, String>;
+        match request.topology.clone() {
+            None => return true,
+            Some(topology) => match topology.node {
+                None => return true,
+                Some(node_topology) => match node_topology {
+                    NodeTopology::Labelled(labelled_topology) => {
+                        // The labels in Volume Node Topology should match the node labels if
+                        // present, otherwise selection of any pool is allowed.
+                        if !labelled_topology.inclusion.is_empty() {
+                            volume_node_topology_inclusion_labels = labelled_topology.inclusion
+                        } else {
+                            return true;
+                        }
+                    }
+                    NodeTopology::Explicit(_) => todo!(),
+                },
+            },
+        };
+
+        // We will reach this part of code only if the volume has inclusion/exclusion labels.
+        match request.registry().specs().node(&item.pool.node) {
+            Ok(spec) => {
+                let inc_match = qualifies_inclusion_labels(
+                    volume_node_topology_inclusion_labels,
+                    spec.labels().clone(),
```

Comment:
Cant we pass references?
```suggestion
                    spec.labels(),
```